File: src/main/java/com/paragon/client/systems/module/impl/movement/Step.java
package com.paragon.client.systems.module.impl.movement;

import com.paragon.api.event.player.StepEvent;
import com.paragon.api.util.string.StringUtil;
import com.paragon.api.module.Module;
import com.paragon.api.module.Category;
import com.paragon.api.setting.Setting;
import com.paragon.asm.mixins.accessor.IMinecraft;
import com.paragon.asm.mixins.accessor.ITimer;
import me.wolfsurge.cerauno.listener.Listener;
import net.minecraft.network.play.client.CPacketPlayer;
import net.minecraftforge.event.entity.living.LivingEvent.LivingUpdateEvent;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
import net.minecraftforge.fml.common.gameevent.TickEvent;

import java.util.HashMap;
import java.util.Map;

/**
 * @author Surge, Doogie13, aesthetical
 * Uses Cosmos's step event, originally from auto which Aestheticall implemented into Cosmos with auto's permission
 */
public class Step extends Module {

    // if you prefer to not use an anonymous inner class, you can always .put in the constructor
    private static final Map<Double, double[]> NCP_STEP_OFFSETS = new HashMap<Double, double[]>() {{
        // note to developers: NCP strict has a "step patch" where it will lag you back on the top of the block once you successfully stepped
        // the bypass for this is to add an extra offset value of the step height. this will not interfere with normal NCP

        put(0.875, new double[] { 0.39, 0.7, 0.875 });
        put(1.0, new double[] { 0.42, 0.75, 1.0 });

        // you can add the rest of ur step positions here
    }};

    public static Step INSTANCE;

    // Step mode
    public static Setting<Mode> mode = new Setting<>("Mode", Mode.NCP)
            .setDescription("What mode to use");

    // Vanilla step height
    public static Setting<Float> stepHeight = new Setting<>("StepHeight", 1.5f, 0.5f, 2.5f, 0.5f)
            .setDescription("How high to step up")
            .setVisibility(() -> mode.getValue().equals(Mode.VANILLA));

    public static Setting<Boolean> useTimer = new Setting<>("UseTimer", true)
            .setDescription("If to use timer to prevent the MORE_PACKETS flag on NCP")
            .setVisibility(() -> mode.getValue().equals(Mode.NCP));

    public Step() {
        super("Step", Category.MOVEMENT, "Lets you instantly step up blocks");

        INSTANCE = this;
    }

    private boolean timer = false;

    @Override
    public void onDisable() {
        if (nullCheck()) {
            return;
        }

        // Set step height to normal
        mc.player.stepHeight = 0.6f;

        // reset our tickLength to 50.0f (1 timer speed)
        ((ITimer) ((IMinecraft) mc).getTimer()).setTickLength(50.0f);
        timer = false;
    }

    @SubscribeEvent
    public void onTick(TickEvent.ClientTickEvent event) {
        if (nullCheck() ||
                // wtf
                !isEnabled()) {
            return;
        }

        // Set step height
        mc.player.stepHeight = stepHeight.getValue();
    }

    @SubscribeEvent
    public void onUpdate(LivingUpdateEvent event) {
        if (event.getEntityLiving().equals(mc.player)) {

            // if we have used timer before and we are on ground after stepping, reset our timer
            if (timer && mc.player.onGround) {
                timer = false;
                ((ITimer) ((IMinecraft) mc).getTimer()).setTickLength(50.0f);
            }
        }
    }

    @Listener
    public void onStep(StepEvent event) {
        if (mode.getValue().equals(Mode.NCP) && event.getEntity().equals(mc.player)) {
            double height = event.getBB().minY - mc.player.posY;

            // don't step if there are any flagging conditions
            if (height > stepHeight.getValue() || !mc.player.onGround || mc.player.isInWater() || mc.player.isInLava()) {
                return;
            }

            // get our packet offsets from the map, and then a default value of null
            double[] offsets = NCP_STEP_OFFSETS.getOrDefault(height, null);
            if (offsets == null || offsets.length == 0) {
                return;
            }

            if (useTimer.getValue()) {

                // set our timer dynamically based off of the amount of offsets we are using
                ((ITimer) ((IMinecraft) mc).getTimer()).setTickLength(50.0f / (1.0f / (offsets.length + 1.0f)));
                timer = true;
            }

            // Send offsets - this simulates a fake jump
            for (double offset : offsets) {
                mc.player.connection.sendPacket(new CPacketPlayer.Position(mc.player.posX, mc.player.posY + offset, mc.player.posZ, false));
            }
        }
    }

    @Override
    public String getData() {
        return StringUtil.getFormattedText(mode.getValue());
    }

    public enum Mode {
        /**
         * Vanilla step - bypasses almost no servers :P
         */
        VANILLA,

        /**
         * NCP step - bypasses the NCP anticheat
         *
         * Is notable that some step heights over 1.5 are not possible on Strict NCP
         */
        NCP
    }

}

File: src/main/kotlin/com/paragon/impl/module/movement/Step.kt
package com.paragon.impl.module.movement

import com.paragon.impl.event.player.StepEvent
import com.paragon.impl.module.Module
import com.paragon.impl.setting.Setting
import com.paragon.bus.listener.Listener
import com.paragon.impl.module.Category
import com.paragon.mixins.accessor.IMinecraft
import com.paragon.mixins.accessor.ITimer
import com.paragon.util.anyNull
import com.paragon.util.string.StringUtil.getFormattedText
import net.minecraft.network.play.client.CPacketPlayer
import net.minecraftforge.event.entity.living.LivingEvent.LivingUpdateEvent
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
import net.minecraftforge.fml.common.gameevent.TickEvent.ClientTickEvent

/**
 * @author Surge, Doogie13, aesthetical
 * Uses Cosmos's step event, originally from auto which Aestheticall implemented into Cosmos with auto's permission
 */
object Step : Module("Step", Category.MOVEMENT, "Lets you instantly step up blocks") {

    // Step mode
    private val mode = Setting("Mode", Mode.NCP) describedBy "What mode to use"

    // Vanilla step height
    private val stepHeight = Setting(
        "StepHeight", 1.5f, 0.5f, 2.5f, 0.5f
    ) describedBy "How high to step up"
    private val useTimer = Setting(
        "UseTimer", true
    ) describedBy "If to use timer to prevent the MORE_PACKETS flag on NCP" visibleWhen { mode.value == Mode.NCP }

    private val ncpOffsets: Map<Double, DoubleArray> = hashMapOf(
        0.875 to doubleArrayOf(0.39, 0.7, 0.875),
        1.0 to doubleArrayOf(0.42, 0.75, 1.0),
        1.5 to doubleArrayOf(0.42, 0.78, 0.63, 0.51, 0.9, 1.21, 1.45, 1.43),
        2.0 to doubleArrayOf(0.425, 0.821, 0.699, 0.599, 1.022, 1.372, 1.652, 1.869, 2.019, 1.919)
    )

    private var timer = false

    override fun onDisable() {
        if (minecraft.anyNull) {
            return
        }

        // Set step height to normal
        minecraft.player.stepHeight = 0.6f

        // reset our tickLength to 50.0f (1 timer speed)
        ((minecraft as IMinecraft).hookGetTimer() as ITimer).hookSetTickLength(50.0f)
        timer = false
    }

    @SubscribeEvent
    fun onTick(event: ClientTickEvent?) {
        if (minecraft.anyNull) {
            return
        }

        // Set step height
        minecraft.player.stepHeight = stepHeight.value
    }

    @SubscribeEvent
    fun onUpdate(event: LivingUpdateEvent) {
        if (event.entityLiving == minecraft.player) {

            // if we have used timer before and we are on ground after stepping, reset our timer
            if (timer && minecraft.player.onGround) {
                timer = false
                ((minecraft as IMinecraft).hookGetTimer() as ITimer).hookSetTickLength(50.0f)
            }
        }
    }

    @Listener
    fun onStep(event: StepEvent) {
        if (mode.value == Mode.NCP && event.entity == minecraft.player && !minecraft.player.capabilities.isFlying) {
            val height: Double = event.bB.minY - minecraft.player.posY

            // don't step if there are any flagging conditions
            if (height > stepHeight.value || !minecraft.player.onGround || minecraft.player.isInWater || minecraft.player.isInLava) {
                return
            }

            // get our packet offsets from the map, and then a default value of null
            val offsets = ncpOffsets.getOrDefault(height, null)

            if (offsets == null || offsets.isEmpty()) {
                return
            }

            if (useTimer.value) {
                // set our timer dynamically based off of the amount of offsets we are using
                ((minecraft as IMinecraft).hookGetTimer() as ITimer).hookSetTickLength(50.0f / (1.0f / (offsets.size + 1.0f)))
                timer = true
            }

            // Send offsets - this simulates a fake jump
            for (offset in offsets) {
                minecraft.player.connection.sendPacket(
                    CPacketPlayer.Position(
                        minecraft.player.posX, minecraft.player.posY + offset, minecraft.player.posZ, false
                    )
                )
            }
        }
    }

    override fun getData(): String {
        return getFormattedText(mode.value)
    }

    @Suppress("UNUSED")
    enum class Mode {
        /**
         * Vanilla step - bypasses almost no servers :P
         */
        VANILLA,

        /**
         * NCP step - bypasses the NCP anticheat
         *
         * Is notable that some step heights over 1.5 are not possible on Strict NCP
         */
        NCP
    }

}